x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/bulk/TransportBulkShardOperationsAction.java

```
@@ -81,73 +115,40 @@ public static CcrWritePrimaryResult shardOperationOnPrimary(
             if (result.getResultType() == Engine.Result.Type.SUCCESS) {
                 assert result.getSeqNo() == targetOp.seqNo();
                 appliedOperations.add(targetOp);
                 location = locationToSync(location, result.getTranslogLocation());
             } else {
                 if (result.getFailure() instanceof AlreadyProcessedFollowingEngineException) {
-                    // Skipped operations will be delivered to replicas via primary-replica resync or peer-recovery.
-                    // The primary must not acknowledge this request until the global checkpoint is at least the highest
-                    // seqno of all skipped operations (i.e., all skipped operations have been processed on every replica).
-                    waitingForGlobalCheckpoint = SequenceNumbers.max(waitingForGlobalCheckpoint, targetOp.seqNo());
+                    // The existing operations below the global checkpoint won't be replicated as they were processed
+                    // in every replicas already. However, the existing operations above the global checkpoint will be
+                    // replicated to replicas but with the existing primary term (not the current primary term) in order
+                    // to guarantee the consistency between the primary and replicas, and between translog and Lucene index.
+                    final AlreadyProcessedFollowingEngineException failure = (AlreadyProcessedFollowingEngineException) result.getFailure();
+                    assert failure.getSeqNo() == targetOp.seqNo() : targetOp.seqNo() + " != " + failure.getSeqNo();
+                    if (failure.getExistingPrimaryTerm().isPresent()) {
+                        appliedOperations.add(rewriteOperationWithPrimaryTerm(sourceOp, failure.getExistingPrimaryTerm().getAsLong()));
+                    } else {
+                        assert targetOp.seqNo() <= primary.getGlobalCheckpoint() : targetOp.seqNo() + " > " + primary.getGlobalCheckpoint();
```

I think this should also throw an exception so replication stops and we'll know about it (assertion is fine for testing too).